Mathlib/Algebra/Polynomial/Degree/IsMonicOfDegree.lean

```diff
@@ -0,0 +1,269 @@
+  · exact natDegree_eq_of_le_of_coeff_ne_zero H₁ <| H₂ ▸ one_ne_zero
+  · exact monic_of_natDegree_le_of_coeff_eq_one n H₁ H₂
+
+lemma IsMonicOfDegree.exists_natDegree_lt {R : Type*} [Semiring R] {p : R[X]} {n : ℕ}
+    (hn : n ≠ 0)  (hp : IsMonicOfDegree p n) :
+    ∃ q : R[X], p = X ^ n + q ∧ q.natDegree < n := by
+  refine ⟨_, hp.natDegree_eq ▸ hp.monic.as_sum, LE.le.trans_lt ?_ (Nat.sub_one_lt hn)⟩
+  refine natDegree_sum_le_of_forall_le _ _ fun i hi ↦ ?_
+  have : i ≤ n - 1 := by rw [Finset.mem_range] at hi; omega
+  compute_degree!
+
+lemma IsMonicOfDegree.natDegree_sub_X_pow {R : Type*} [Ring R] {p : R[X]} {n : ℕ} (hn : n ≠ 0)
+    (hp : IsMonicOfDegree p n) :
+    (p - X ^ n).natDegree < n := by
+  obtain ⟨q, hq₁, hq₂⟩ := hp.exists_natDegree_lt hn
+  simpa [hq₁]
+
+lemma IsMonicOfDegree.natDegree_sub_lt {R : Type*} [Ring R] {p q : R[X]} {n : ℕ} (hn : n ≠ 0)
+    (hp : IsMonicOfDegree p n) (hq : IsMonicOfDegree q n) :
+    (p - q).natDegree < n := by
+  rw [← sub_sub_sub_cancel_right p q (X ^ n)]
+  replace hp := hp.natDegree_sub_X_pow hn
+  replace hq := hq.natDegree_sub_X_pow hn
+  set p' := p - X ^ n -- do not confuse `compute_degree!`
+  set q' := q - X ^ n
+  compute_degree!
+
+lemma IsMonicOfDegree.mul {R : Type*} [Semiring R] {p q : R[X]} {m n : ℕ}
+    (hp : IsMonicOfDegree p m) (hq : IsMonicOfDegree q n) :
+    IsMonicOfDegree (p * q) (m + n) := by
+  rcases subsingleton_or_nontrivial R with H | H
+  · simp only [isMonicOfDegree_iff_of_subsingleton, Nat.add_eq_zero] at hp hq ⊢
+    exact ⟨hp, hq⟩
+  refine ⟨?_, hp.2.mul hq.2⟩
+  have : p.leadingCoeff * q.leadingCoeff ≠ 0 := by
+    rw [hp.leadingCoeff_eq, hq.leadingCoeff_eq, one_mul]
+    exact one_ne_zero
+  rw [natDegree_mul' this, hp.1, hq.1]
+
+lemma IsMonicOfDegree.pow {R : Type*} [Semiring R] {p : R[X]} {m : ℕ} (hp : IsMonicOfDegree p m)
+    (n : ℕ) :
+    IsMonicOfDegree (p ^ n) (m * n) := by
+  induction n with
+  | zero => simp
+  | succ n ih =>
+    rw [pow_succ, mul_add, mul_one]
+    exact ih.mul hp
+
+lemma isMonicOfDegree_X (R : Type*) [Semiring R] [Nontrivial R] : IsMonicOfDegree (X : R[X]) 1 :=
+  (isMonicOfDegree_iff ..).mpr ⟨natDegree_X_le, coeff_X_one⟩
+
+lemma isMonicOfDegree_X_pow (R : Type*) [Semiring R] [Nontrivial R] (n : ℕ) :
+    IsMonicOfDegree ((X : R[X]) ^ n) n :=
+  (isMonicOfDegree_iff ..).mpr ⟨natDegree_X_pow_le n, coeff_X_pow_self n⟩
+
+lemma IsMonicOfDegree.coeff_eq {R : Type*} [Semiring R] {p q : R[X]} {n : ℕ}
+    (hp : IsMonicOfDegree p n) (hq : IsMonicOfDegree q n) {m : ℕ} (hm : n ≤ m) :
+    p.coeff m = q.coeff m := by
+  nontriviality R
+  rw [isMonicOfDegree_iff] at hp hq
+  rcases eq_or_lt_of_le hm with rfl | hm
+  · rw [hp.2, hq.2]
+  · replace hp : p.natDegree < m := hp.1.trans_lt hm
+    replace hq : q.natDegree < m := hq.1.trans_lt hm
+    rw [coeff_eq_zero_of_natDegree_lt hp, coeff_eq_zero_of_natDegree_lt hq]
+
+lemma IsMonicOfDegree.of_mul_left {R : Type*} [Semiring R] {p q : R[X]} {m n : ℕ}
```

Comment:
Is this an `iff` between `hpq` and the current conclusion?

Reply:
It is really that any two of `IsMonicOfDegree p m`, `IsMonicOfDegree q n` and `IsMonicOfDegree (p * q) (m + n)` imply the third (the three cases are `.mul`, `.of_mul_left` and `.of_mul_right`). I don't think there is a convenient way to express this.